### DISCUSSION TITLE
Sorting results consistency

### DISCUSSION THREAD

**Discussion by roi-tevel** [JanusGraph/janusgraph#3042] [Q&A] [answered]
Hi, 
Assuming we have a query sorted by a key where the key values of some of the vertices are the same, does Janus graph ensures results consistency? When I try it seems to always sort by ids if sorting values are the same, but I am asking to make sure this is indeed always so. 
Thanks, 
roi

**Comment by li-boxuan** [comment, marked as answer]
It depends on your storage backend. For example, if you are using Cassandra then the order is deterministic. JanusGraph itself does not guarantee so.

**Reply by roi-tevel** [reply]
thanks, is it also deterministic if I am using Bigtable? (and is there a way to find out if a storage backend guarantee so?)

**Reply by li-boxuan** [reply]
There are three cases:
1. If your query hits an Elasticsearch index, then the result order would always be non-deterministic no matter what storage backend you use.
2. If your query does not hit any index, then the result order depends on whether the storage backend stores the results in a consistent order. Since Bigtable [stores data lexicographically by row key](https://cloud.google.com/bigtable/docs/schema-design), the result order would be deterministic.
3. If your query hits a composite index, then the result order depends on whether the storage backend supports a clustering key/sort key. BigTable, Cassandra, and HBase all support it. Thus, the result order would be deterministic.
